File: strStr.py
# Time Complexity : O(n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if needle not in haystack and len(needle)>len(haystack):
            return -1
        i = 0
        n = len(needle)
        while i < len(haystack) and n <= len(haystack):
            val = slice(i,n)
            if haystack[val] == needle:
                return i
            i += 1
            n += 1
        return -1

File: Strstr.java
// TC = O(mn)
// SC = O(1)
// Here we are using slinding window approch.
class Solution {
    public int strStr(String haystack, String needle) {
        int m = needle.length();
        int n = haystack.length();
        
        for(int wind = 0; wind<=n-m; wind++){
            for(int i=0; i<m; i++){
                if(needle.charAt(i) != haystack.charAt(wind+i)){
                    break;
                }
                if(i==m-1){
                    return wind;
                }
            }
        }
        return -1;
    }
}